File: hooks/scripts/forget-cleanup.py
#!/usr/bin/env python3
"""Cleanup engine for claude-forget. Reads the pending flag and the hook payload
and, when safe, performs the armed wipe.

  scope=session  → delete this session's transcript + project memory
  scope=project  → delete every Claude-realm reference to the project

Inputs via env: FORGET_FLAG, FORGET_CLAUDE_DIR, FORGET_PAYLOAD, FORGET_DRY_RUN.
Set FORGET_DRY_RUN=1 to print planned actions instead of performing them.
"""
import json
import os
import shutil
import sys
import tempfile
import time

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import forget_common as fc

FLAG       = os.environ["FORGET_FLAG"]
CLAUDE_DIR = os.environ["FORGET_CLAUDE_DIR"]
PAYLOAD    = os.environ.get("FORGET_PAYLOAD", "")
DRY        = os.environ.get("FORGET_DRY_RUN", "0") == "1"

CLAUDE_JSON = os.path.expanduser("~/.claude.json")
HISTORY     = os.path.join(CLAUDE_DIR, "history.jsonl")
SECLOG      = os.path.join(CLAUDE_DIR, "security", "log.txt")
BACKUPS     = os.path.join(CLAUDE_DIR, "backups")
SESSIONS    = os.path.join(CLAUDE_DIR, "sessions")


def done():
    raise SystemExit(0)


def rp(p):
    try:
        return os.path.realpath(p)
    except Exception:
        return p


def rm(path, label):
    if not path:
        return
    if DRY:
        exists = os.path.exists(path) or os.path.islink(path)
        kind = "dir " if os.path.isdir(path) and not os.path.islink(path) else "file"
        print(f"[dry-run] {'DELETE' if exists else 'skip  '} {kind} {label}: {path}")
        return
    try:
        if os.path.isdir(path) and not os.path.islink(path):
            shutil.rmtree(path)
        elif os.path.exists(path) or os.path.islink(path):
            os.remove(path)
    except FileNotFoundError:
        pass
    except Exception:
        pass


def atomic_write(path, data):
    d = os.path.dirname(path) or "."
    fd, tmp = tempfile.mkstemp(dir=d)
    with os.fdopen(fd, "w") as out:
        out.write(data)
    os.replace(tmp, path)


def scrub_jsonl(path, drop_pred, label):
    """Drop JSONL lines for which drop_pred(parsed_obj) is True."""
    if not os.path.isfile(path):
        return
    try:
        kept, dropped = [], 0
        with open(path, errors="replace") as fh:
            for ln in fh:
                s = ln.strip()
                if not s:
                    continue
                try:
                    obj = json.loads(s)
                except Exception:
                    kept.append(ln if ln.endswith("\n") else ln + "\n")
                    continue
                if drop_pred(obj):
                    dropped += 1
                    continue
                kept.append(ln if ln.endswith("\n") else ln + "\n")
        if DRY:
            print(f"[dry-run] scrub {label}: drop {dropped} line(s)")
            return
        atomic_write(path, "".join(kept))
    except Exception:
        pass


def scrub_text_lines(path, contains, label):
    if not os.path.isfile(path):
        return
    try:
        kept, dropped = [], 0
        with open(path, errors="replace") as fh:
            for ln in fh:
                if contains in ln:
                    dropped += 1
                    continue
                kept.append(ln)
        if DRY:
            print(f"[dry-run] scrub {label}: drop {dropped} line(s)")
            return
        atomic_write(path, "".join(kept))
    except Exception:
        pass


def remove_project_from_json(path, cwd, label, backup=False):
    """Remove projects[cwd] from a .claude.json-style file (re-read fresh)."""
    if not os.path.isfile(path):
        return
    try:
        d = json.load(open(path))
    except Exception:
        return
    if not (isinstance(d.get("projects"), dict) and cwd in d["projects"]):
        return
    if DRY:
        print(f"[dry-run] {label}: remove projects[\"{cwd}\"]"
              + (" (after backup)" if backup else ""))
        return
    try:
        if backup:
            shutil.copy2(path, os.path.join(BACKUPS, f".claude.json.prenuke.{int(time.time()*1000)}"))
        del d["projects"][cwd]
        atomic_write(path, json.dumps(d, indent=2))
    except Exception:
        pass


# ---- load flag + payload ---------------------------------------------------
if not os.path.isfile(FLAG):
    done()
try:
    flag = json.load(open(FLAG))
except Exception:
    done()
try:
    ev = json.loads(PAYLOAD) if PAYLOAD.strip() else {}
except Exception:
    ev = {}

event  = ev.get("hook_event_name", "")
source = ev.get("source", "")
ev_cwd = ev.get("cwd", "") or os.getcwd()
ev_tx  = ev.get("transcript_path", "")
ev_sid = ev.get("session_id", "")

scope  = flag.get("scope", "session")
f_tx   = flag.get("transcript", "")
f_sid  = flag.get("session_id", "")
f_cwd  = flag.get("cwd", "")

# Only act in the project the wipe was armed in.
if f_cwd and ev_cwd and rp(f_cwd) != rp(ev_cwd):
    done()

same_as_active = bool(ev_tx) and rp(ev_tx) == rp(f_tx)


def drop_flag():
    if not DRY:
        try:
            os.remove(FLAG)
        except Exception:
            pass
    else:
        print("[dry-run] would remove flag:", FLAG)


def update_flag(extra):
    """Keep the flag but merge in extra fields (so the next SessionStart can
    finish cleaning up — e.g. the one-line ghost transcript /exit leaves behind)."""
    if DRY:
        print("[dry-run] keep flag, set:", extra)
        return
    try:
        d = json.load(open(FLAG))
        d.update(extra)
        atomic_write(FLAG, json.dumps(d))
    except Exception:
        pass


# ===========================================================================
# SESSION SCOPE
# ===========================================================================
if scope == "session":
    f_mem = flag.get("memory_dir", "")
    # finalize=True  → remove the flag (job done, stable)
    # finalize=False → keep the flag so the NEXT SessionStart re-runs and clears
    #                  the one-line ghost transcript that /exit recreates post-hook.
    if event == "SessionStart":
        if source == "compact":
            done()                       # same session continues mid-flight
        if same_as_active:
            if source == "resume":
                drop_flag()              # resumed the armed session → cancel
            done()
        finalize = True                  # startup / clear / resume-other → armed is closed
    elif event == "SessionEnd":
        if ev_sid and f_sid and ev_sid != f_sid:
            done()
        if (not ev_sid) and ev_tx and not same_as_active:
            done()
        finalize = False                 # delete now, but keep flag for the ghost sweep
    else:
        done()

    sid = f_sid or (os.path.splitext(os.path.basename(f_tx))[0] if f_tx else "")
    rm(f_tx, "transcript")
    rm(f_tx + ".meta", "transcript meta")
    if f_tx.endswith(".jsonl"):
        rm(f_tx[:-len(".jsonl")], "session subdir")
    if sid:
        rm(os.path.join(CLAUDE_DIR, "session-env", sid), "session-env")
        scrub_jsonl(HISTORY, lambda o: o.get("sessionId") == sid, "history.jsonl")
    rm(f_mem, "project memory")
    drop_flag() if finalize else update_flag({"ghost_sweep": True})
    done()


# ===========================================================================
# PROJECT SCOPE
# ===========================================================================
project_dir = flag.get("project_dir", "")
cwd = f_cwd
heavy_done = flag.get("heavy_done", False)

# Decide whether (and how) to fire.
preserve_sid = None
preserve_tx = None
finalize = True
if event == "SessionEnd":
    if not ((f_sid and ev_sid and ev_sid == f_sid)
            or (ev_tx and rp(ev_tx) == rp(f_tx))
            or (not ev_sid and not ev_tx)):
        done()  # a different session ended; not the one that armed the nuke
    finalize = False  # nuke now, but keep flag so the next start sweeps the ghost
elif event == "SessionStart":
    if source == "resume" and same_as_active:
        drop_flag()  # user resumed the armed session → cancel the nuke
        done()
    if source in ("startup", "resume"):
        preserve_sid, preserve_tx = ev_sid, ev_tx  # keep the new session
    else:
        done()  # /clear and compact do NOT complete a project nuke
else:
    done()

# On the follow-up ghost sweep (SessionEnd already did the heavy global scrubs)
# only clean leftover transcripts/per-session refs — do NOT re-scrub global config,
# so the freshly-reopened project's new ~/.claude.json entry is left intact.
do_globals = not heavy_done

ids = fc.project_session_ids(project_dir, HISTORY, cwd)
if preserve_sid:
    ids.discard(preserve_sid)

# 1) the project's session/memory directory
if preserve_tx and project_dir and rp(os.path.dirname(preserve_tx)) == rp(project_dir):
    keep_base = os.path.basename(preserve_tx)
    keep_sub = keep_base[:-len(".jsonl")] if keep_base.endswith(".jsonl") else None
    for n in (os.listdir(project_dir) if os.path.isdir(project_dir) else []):
        if n == keep_base or (keep_sub and n == keep_sub):
            continue
        rm(os.path.join(project_dir, n), "project entry")
else:
    rm(project_dir, "project dir")

# 2) per-session references for every project session id
for i in sorted(ids):
    for p in fc.per_session_paths(CLAUDE_DIR, i):
        rm(p, f"per-session ({i[:8]})")

# 3) sessions/*.json (PID→cwd map) pointing at this project
if os.path.isdir(SESSIONS):
    for n in os.listdir(SESSIONS):
        p = os.path.join(SESSIONS, n)
        try:
            txt = open(p, errors="replace").read()
        except Exception:
            continue
        if cwd in txt and not (preserve_sid and preserve_sid in txt):
            rm(p, "sessions map")

if do_globals:
    # 4) prompt history lines for this project (keep the preserved session's, if any)
    scrub_jsonl(HISTORY,
                lambda o: o.get("project") == cwd and o.get("sessionId") != preserve_sid,
                "history.jsonl")

    # 5) ~/.claude.json project entry (best-effort; backed up first)
    remove_project_from_json(CLAUDE_JSON, cwd, "~/.claude.json", backup=True)

    # 6) security audit log lines mentioning the path
    scrub_text_lines(SECLOG, cwd, "security/log.txt")

    # 7) strip the project entry from rotating ~/.claude.json backups
    for bp in fc.backups_with_project(BACKUPS, cwd):
        remove_project_from_json(bp, cwd, f"backup {os.path.basename(bp)}")

drop_flag() if finalize else update_flag({"heavy_done": True})
done()

File: hooks/forget-cleanup.mjs
#!/usr/bin/env node
// forget-cleanup.mjs — hook entrypoint for SessionStart and SessionEnd.
// Iterates EVERY armed flag (one per session) and performs the ones that are safe
// for this event. Reads the hook payload on stdin.
//   scope=session → delete that session's transcript + project memory
//   scope=project → delete every Claude-realm reference to the project
// Set FORGET_DRY_RUN=1 to print planned actions instead of performing them.
import fs from "node:fs";
import path from "node:path";
import * as fc from "./forget-common.mjs";

const CLAUDE_DIR = fc.CLAUDE_DIR;
const DRY = process.env.FORGET_DRY_RUN === "1";

const CLAUDE_JSON = fc.CLAUDE_JSON;
const HISTORY = path.join(CLAUDE_DIR, "history.jsonl");
const SECLOG = path.join(CLAUDE_DIR, "security", "log.txt");
const BACKUPS = path.join(CLAUDE_DIR, "backups");
const SESSIONS = path.join(CLAUDE_DIR, "sessions");

const log = (s) => process.stdout.write(s + "\n");
const exit = () => process.exit(0);

// Fast exit: nothing armed.
const flags = fc.listPendingFlags();
if (flags.length === 0) exit();

let payload = "";
try { payload = fs.readFileSync(0, "utf8"); } catch { payload = ""; }
let ev = {};
if (payload.trim()) { try { ev = JSON.parse(payload); } catch { ev = {}; } }

const event = ev.hook_event_name || "";
const source = ev.source || "";
const evCwd = ev.cwd || process.cwd();
const evTx = ev.transcript_path || "";
const evSid = ev.session_id || "";

const rp = (p) => { try { return fs.realpathSync(p); } catch { return p; } };

function rm(p, label) {
  if (!p) return;
  if (DRY) {
    const kind = fc.isDir(p) ? "dir " : "file";
    log(`[dry-run] ${fc.exists(p) ? "DELETE" : "skip  "} ${kind} ${label}: ${p}`);
    return;
  }
  try { fs.rmSync(p, { recursive: true, force: true }); } catch { /* ignore */ }
}

function atomicWrite(p, data) {
  const dir = path.dirname(p) || ".";
  const tmp = path.join(dir, `.forget-tmp-${process.pid}-${Date.now()}`);
  fs.writeFileSync(tmp, data);
  fs.renameSync(tmp, p);
}

function scrubJsonl(p, dropPred, label) {
  if (!fs.existsSync(p)) return;
  try {
    const kept = [];
    let dropped = 0;
    for (const ln of fs.readFileSync(p, "utf8").split("\n")) {
      const s = ln.trim();
      if (!s) continue;
      let obj;
      try { obj = JSON.parse(s); } catch { kept.push(ln); continue; }
      if (dropPred(obj)) { dropped++; continue; }
      kept.push(ln);
    }
    if (DRY) { log(`[dry-run] scrub ${label}: drop ${dropped} line(s)`); return; }
    atomicWrite(p, kept.length ? kept.join("\n") + "\n" : "");
  } catch { /* ignore */ }
}

function scrubTextLines(p, contains, label) {
  if (!fs.existsSync(p)) return;
  try {
    const kept = [];
    let dropped = 0;
    for (const ln of fs.readFileSync(p, "utf8").split("\n")) {
      if (ln.includes(contains)) { dropped++; continue; }
      kept.push(ln);
    }
    if (DRY) { log(`[dry-run] scrub ${label}: drop ${dropped} line(s)`); return; }
    atomicWrite(p, kept.join("\n"));
  } catch { /* ignore */ }
}

function removeProjectFromJson(p, cwd, label, backup = false) {
  if (!fs.existsSync(p)) return;
  const d = fc.readJSON(p);
  if (!d || !d.projects || typeof d.projects !== "object" ||
      !Object.prototype.hasOwnProperty.call(d.projects, cwd)) return;
  if (DRY) { log(`[dry-run] ${label}: remove projects["${cwd}"]${backup ? " (after backup)" : ""}`); return; }
  try {
    if (backup) {
      try { fs.copyFileSync(p, path.join(BACKUPS, `.claude.json.prenuke.${Date.now()}`)); } catch { /* ignore */ }
    }
    delete d.projects[cwd];
    atomicWrite(p, JSON.stringify(d, null, 2));
  } catch { /* ignore */ }
}

function dropFlag(flagPath) {
  if (DRY) { log(`[dry-run] would remove flag: ${flagPath}`); return; }
  try { fs.rmSync(flagPath, { force: true }); } catch { /* ignore */ }
}
function updateFlag(flagPath, extra) {
  if (DRY) { log(`[dry-run] keep flag ${path.basename(flagPath)}, set: ${JSON.stringify(extra)}`); return; }
  try { const d = fc.readJSON(flagPath) || {}; Object.assign(d, extra); atomicWrite(flagPath, JSON.stringify(d)); } catch { /* ignore */ }
}

// Process one armed flag against the current hook event.
function processFlag(flagPath, flag) {
  const scope = flag.scope || "session";
  const fTx = flag.transcript || "";
  const fSid = flag.session_id || "";
  const fCwd = flag.cwd || "";

  // Only act in the project this flag was armed in.
  if (fCwd && evCwd && rp(fCwd) !== rp(evCwd)) return;

  const sameAsActive = !!evTx && rp(evTx) === rp(fTx);

  // ---- SESSION SCOPE ----
  if (scope === "session") {
    const fMem = flag.memory_dir || "";
    let finalize = true;
    if (event === "SessionStart") {
      if (source === "compact") return;          // same session continues
      if (sameAsActive) {
        if (source === "resume") dropFlag(flagPath); // resumed armed session → cancel
        return;
      }
      finalize = true;                            // startup / clear / resume-other → armed is closed
    } else if (event === "SessionEnd") {
      if (evSid && fSid && evSid !== fSid) return;
      if (!evSid && evTx && !sameAsActive) return;
      finalize = false;                           // delete now, keep flag for the ghost sweep
    } else {
      return;
    }

    const sid = fSid || (fTx ? path.basename(fTx).replace(/\.jsonl$/, "") : "");
    rm(fTx, "transcript");
    rm(fTx + ".meta", "transcript meta");
    if (fTx.endsWith(".jsonl")) rm(fTx.slice(0, -".jsonl".length), "session subdir");
    if (sid) {
      rm(path.join(CLAUDE_DIR, "session-env", sid), "session-env");
      scrubJsonl(HISTORY, (o) => o.sessionId === sid, "history.jsonl");
    }
    rm(fMem, "project memory");
    if (finalize) dropFlag(flagPath); else updateFlag(flagPath, { ghost_sweep: true });
    return;
  }

  // ---- PROJECT SCOPE ----
  const projectDir = flag.project_dir || "";
  const cwd = fCwd;
  const heavyDone = !!flag.heavy_done;

  let preserveSid = null;
  let preserveTx = null;
  let finalize = true;

  if (event === "SessionEnd") {
    if (!((fSid && evSid && evSid === fSid) ||
          (evTx && rp(evTx) === rp(fTx)) ||
          (!evSid && !evTx))) return;             // a different session ended
    finalize = false;                             // nuke now, keep flag for the ghost sweep
  } else if (event === "SessionStart") {
    if (source === "resume" && sameAsActive) { dropFlag(flagPath); return; } // resumed armed → cancel
    if (source === "startup" || source === "resume") {
      preserveSid = evSid; preserveTx = evTx;     // keep the new session
      finalize = true;
    } else {
      return;                                     // /clear and compact do NOT complete a project nuke
    }
  } else {
    return;
  }

  const doGlobals = !heavyDone;

  const ids = fc.projectSessionIds(projectDir, HISTORY, cwd);
  if (preserveSid) ids.delete(preserveSid);

  // 1) project session/memory directory
  if (preserveTx && projectDir && rp(path.dirname(preserveTx)) === rp(projectDir)) {
    const keepBase = path.basename(preserveTx);
    const keepSub = keepBase.endsWith(".jsonl") ? keepBase.slice(0, -".jsonl".length) : null;
    for (const n of fc.listDir(projectDir)) {
      if (n === keepBase || (keepSub && n === keepSub)) continue;
      rm(path.join(projectDir, n), "project entry");
    }
  } else {
    rm(projectDir, "project dir");
  }

  // 2) per-session references for every project session id
  for (const i of [...ids].sort()) {
    for (const p of fc.perSessionPaths(CLAUDE_DIR, i)) rm(p, `per-session (${i.slice(0, 8)})`);
  }

  // 3) sessions/*.json (PID→cwd map) pointing at this project
  if (fc.isDir(SESSIONS)) {
    for (const n of fc.listDir(SESSIONS)) {
      const p = path.join(SESSIONS, n);
      let txt = "";
      try { txt = fs.readFileSync(p, "utf8"); } catch { continue; }
      if (txt.includes(cwd) && !(preserveSid && txt.includes(preserveSid))) rm(p, "sessions map");
    }
  }

  if (doGlobals) {
    // 4) prompt history lines for this project (keep the preserved session's, if any)
    scrubJsonl(HISTORY, (o) => o.project === cwd && o.sessionId !== preserveSid, "history.jsonl");
    // 5) ~/.claude.json project entry (best-effort; backed up first)
    removeProjectFromJson(CLAUDE_JSON, cwd, "~/.claude.json", true);
    // 6) security audit log lines mentioning the path
    scrubTextLines(SECLOG, cwd, "security/log.txt");
    // 7) strip the project entry from rotating ~/.claude.json backups
    for (const bp of fc.backupsWithProject(BACKUPS, cwd)) removeProjectFromJson(bp, cwd, `backup ${path.basename(bp)}`);
  }

  if (finalize) dropFlag(flagPath); else updateFlag(flagPath, { heavy_done: true });
}

for (const { path: flagPath, data: flag } of flags) {
  try { processFlag(flagPath, flag); } catch { /* one bad flag shouldn't block the rest */ }
}
exit();
